### ISSUE TITLE
uri module always following redirects

### ISSUE THREAD

**Issue by rrtj3** [ansible/ansible#5742]
It looks like this pull request:

https://github.com/ansible/ansible/commit/5c0daf9e0a3a58253df61e56e39333637202cd12

is unconditionally setting 'follow_redirects' to True which should not be the case.

**Comment by srgvg** [comment]
Trying to remember what I did here..

The original code let http.follow_redirects to default, which according to http://httplib2.googlecode.com/hg/doc/html/libhttplib2.html#http-objects is 'True'.

My patch just made some preliminary stuff, using http.follow_redirects set to False, then puts it back to True, the original default. Perhaps that default is a problem, but at first look, I don't see which problem this could have introduced?

Why should this default not be the case?

**Comment by rrtj3** [comment]
Hmm, true. I'm wondering if there's some weird interaction between 'follow_redirects' and 'follow_all_redirects' here. e.g. follow_redirects is possibly over-riding follow_all_redirects, or the toggling of follow_redirects is changing 'follow_all_redirects' value.... will have to do some testing/debugging here. Thanks @sergevanginderachter for taking a look.

**Comment by srgvg** [comment]
No problem, still wondering what's your exact problem though?

**Comment by rrtj3** [comment]
User on mailing list is reporting:

"I'm trying to get ansible to submit an http request for specific a url, and verify it gets a 301.  It does not need to follow the location header and make the next http request.  

I figured that follow_redirects=no would solve this for me, however I see ansible make the initial http request, receive a 301, then make the next http request and get a 200 back.  It then fails because I've said status_code=301, it must only care about the second http request it made."

**Comment by rrtj3** [comment]
After some testing. It appears that the only way to have the httplib2 module not follow redirects, 301's at least, is by having both 'follow_all_redirects' and 'follow_redirects' set to false. Here is my simple test code:

``` python
#!/usr/bin/python
import httplib2

url  = "http://www.law.hawaii.edu/library"
method = "GET"

# Create a Http object and set some default options.
h = httplib2.Http(disable_ssl_certificate_validation=True)

# Redirects
h.follow_all_redirects = False
h.follow_redirects = False

# Make the request
resp, content = h.request(url, method=method)
print resp, content
```

so I'll submit a pull request that sets them both to false if the follow_redirection module option is set to no.

**Comment by rrtj3** [comment]
Ok, this is a little more complicated than I initially thought. When I initially wrote the module I didn't fully understand the httplib2 libraries 'follow_redirects' and 'follow_all_redirects' options and how they interact. At this point, the uri modules, follow_redirects option allows the module to follow "unsafe" redirects by setting it's value to "yes" but doesn't disable redirects by setting it's value to "no". (That's where the httplib2's 'follow_redirects' option comes into play.) 

So, long story short, I can either add a new option to the module to explicitly disable following redirects, or I can make the not following safe redirects the default, which would likely break some users tasks that may implicitly rely on this behavior. I'm leaning more towards adding a new option to explicitly disable redirects.  Opinions?

**Comment by bcoca** [comment]
my 2c (probably worth 1/10th) let yes and no function as is and change it
from a boolean to a 'choices' [all, safe, none] with 'yes/true' and
'no/false' being aliases. Make note of old behavior deprecation in docs and
eventually remove it in a version or 2.

**Comment by rrtj3** [comment]
Hmm, I like that a lot. Graceful. I'll wait to see if others have any other opinions before proceeding.

**Comment by rrtj3** [comment]
I'm just gonna go ahead with making the changes as suggested by @bcoca .

**Comment by rrtj3** [comment]
I've sent a pull request which fixes the following of redirects, while keeping backwards compatibility as @bcoca suggested above. I've tested and it works well for me. Any additional testing and/or comments on the commit are welcome.

**Comment by rrtj3** [comment]
Closing this issue since the pull request fixing it was merged.
